src/main/scala/mesosphere/marathon/MarathonModule.scala

```diff
@@ -105,6 +107,41 @@ class MarathonModule(conf: MarathonConf, http: HttpConf, zk: ZooKeeperClient)
 
   @Provides
   @Singleton
+  def provideMarathonSchedulerService(
+    zk: ZooKeeperClient,
+    leadershipCoordinator: LeadershipCoordinator,
+    healthCheckManager: HealthCheckManager,
+    @Named(ModuleNames.CANDIDATE) candidate: Option[Candidate],
+    config: MarathonConf,
+    frameworkIdUtil: FrameworkIdUtil,
+    @Named(ModuleNames.LEADER_ATOMIC_BOOLEAN) leader: AtomicBoolean,
+    appRepository: AppRepository,
+    taskTracker: TaskTracker,
+    driverFactory: SchedulerDriverFactory,
+    system: ActorSystem,
+    migration: Migration,
+    @Named("schedulerActor") schedulerActor: ActorRef,
+    @Named(EventModule.busName) eventStream: EventStream,
+    leadershipCallbacks: Seq[LeadershipCallback] = Seq.empty): MarathonSchedulerService = {
+
+    val service = new MarathonSchedulerService(leadershipCoordinator, healthCheckManager, candidate, config,
+      frameworkIdUtil, leader, appRepository, taskTracker, driverFactory, system, migration, schedulerActor,
+      eventStream, leadershipCallbacks)
+
+    // The current candidate implementation does not handle ZK connection loss
+    // We register a listener on the candidate zk connection and handle disconnetions explicitly
+    zk.get().register(new Watcher {
```

Comment:
It would be nice to move this out of the Guice logic.

It also looks strange to me that you couple this with creating the MarathonSchedulerService.

Couldn't that simply be inside a `WhenLeaderActor` that has the `ZooKeeperClient` and the `MarathonSchedulerService` as dependencies?